src/client/clientmap.cpp

```diff
@@ -180,17 +202,403 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+
+
+			// Calculate the coordinates for range and frutum culling
+			v3f mesh_sphere_center;
+			f32 mesh_sphere_radius;
+
+			v3s16 block_pos_nodes = block_coord * MAP_BLOCKSIZE;
+
+			if (mesh) {
+				mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+						+ mesh->getBoundingSphereCenter();
+				mesh_sphere_radius = mesh->getBoundingRadius();
+			}
+			else {
+				mesh_sphere_center = intToFloat(block_pos_nodes, BS) + v3f((MAP_BLOCKSIZE * 0.5f - 0.5f) * BS);
+				mesh_sphere_radius = 0.0f;
+			}
+
+			// First, perform a simple distance check.
+			if (!m_control.range_all &&
+				mesh_sphere_center.getDistanceFrom(intToFloat(cam_pos_nodes, BS)) >
+					m_control.wanted_range * BS + mesh_sphere_radius)
+				continue; // Out of range, skip.
+
+			// Frustum culling
+			// Only do coarse culling here, to account for fast camera movement.
+			// This is needed because this function is not called every frame.
+			float frustum_cull_extra_radius = 300.0f;
+			if (is_frustum_culled(mesh_sphere_center,
+					mesh_sphere_radius + frustum_cull_extra_radius))
+				continue;
+
+			// Calculate the vector from the camera block to the current block
+			// We use it to determine through which sides of the current block we can continue the search
+			v3s16 look = block_coord - camera_block;
+
+			// Occluded near sides will further occlude the far sides
+			u8 visible_outer_sides = flags & 0x07;
+
+			// Raytraced occlusion culling - send rays from the camera to the block's corners
+			if (occlusion_culling_enabled && m_enable_raytraced_culling &&
+					block && mesh &&
+					visible_outer_sides != 0x07 && isBlockOccluded(block, cam_pos_nodes)) {
+				blocks_occlusion_culled++;
+				continue;
+			}
+
+			// The block is visible, add to the draw list
+			if (mesh) {
+				// Add to set
+				block->refGrab();
+				m_drawlist[block_coord] = block;
+			}
+
+			// Decide which sides to traverse next or to block away
+
+			// First, find the near sides that would occlude the far sides
+			// * A near side can itself be occluded by a nearby block (the test above ^^)
+			// * A near side can be visible but fully opaque by itself (e.g. ground at the 0 level)
+
+			// mesh solid sides are +Z-Z+Y-Y+X-X
+			// do not test sides where we are 'inside' the block's coordinates
+			u8 ignore_inner_sides = (look.X == 0 ? 3 : 0) |
+				(look.Y == 0 ? 12 : 0) |
+				(look.Z == 0 ? 48 : 0);
+
+			// get the mask for the sides that are relevant based on the direction
+			u8 near_inner_sides = (look.X > 0 ? 1 : 2) |
+					(look.Y > 0 ? 4 : 8) |
+					(look.Z > 0 ? 16 : 32);
+			
+			// This bitset is +Z-Z+Y-Y+X-X (See MapBlockMesh), and axis is XYZ.
+			u8 transparent_sides = (occlusion_culling_enabled && block) ? ~block->solid_sides : 0x3F;
+
+			// compress block transparent sides to ZYX mask of see-through axes
+			u8 near_transparency = ((transparent_sides & near_inner_sides) | ignore_inner_sides) & 0x3F;
+
+			near_transparency |= (near_transparency >> 1);
```

Comment:
https://github.com/minetest/minetest/pull/13104#discussion_r1060213082
> The mask is not needed, because we hand-pick the bits later.

If you don't use the mask, you'll shift the bits into different axes.

Isn't what you wanted to do here `X = -X || +X` via `-X = -X || +X; X = -X`?

```suggestion
			near_transparency |= (near_transparency >> 1) & 0x3F;
```

Reply:
The following lines hand-pick bits 0x01, 0x04 and 0x10 (negative sides of the three axes), which together are the same as the mask 0x15 that I removed. It does not matter if the other bits are pollutedm as they are ignored.

Reply:
Ah I see now, ok.

<!---
The important part is that the AND happens *before* the OR. (Your old code did this wrong.)

Example:
Given: `near_transparency = 0b000100 = 0x4` (-Y only)
Resulting `near_transparency` should be `0b010`, right?
Without mask: after this line: `near_transparency = 0b000110` (-Y | +X). After hand-picking: 
-->